src/main/scala/chisel3/util/LoadMemoryTransform.scala

```diff
@@ -0,0 +1,234 @@
+// See LICENSE for license details.
+
+package chisel3.util
+
+import chisel3.core.RunFirrtlTransform
+import chisel3.internal.{Builder, InstanceId}
+import firrtl.annotations._
+import firrtl.ir.{Module => _, _}
+import firrtl.passes.Pass
+import firrtl.passes.memlib.DefAnnotatedMemory
+import firrtl.transforms.BlackBoxInlineAnno
+import firrtl.{CircuitForm, CircuitState, LowForm, Transform, VerilogEmitter, WDefInstance}
+
+import scala.collection.mutable
+import scala.util.matching.Regex
+
+/**
+  * chisel implementation for load memory
+  * @param target        memory to load
+  * @param fileName      name of input file
+  * @param hexOrBinary   use $readmemh or $readmemb
+  */case class ChiselLoadMemoryAnnotation(target: InstanceId, fileName: String, hexOrBinary: String = "h")
+  extends chisel3.core.ChiselAnnotation
+    with RunFirrtlTransform {
+
+  if(! Seq("h","b").contains(hexOrBinary)) {
+    Builder.error(
+      s"""LoadMemory from file $fileName format must be "h" or "b" not "$hexOrBinary" """
+    )
+  }
+
+  if(fileName.isEmpty) {
+    Builder.warning(
+      s"""LoadMemory from file annotations file empty file name"""
+    )
+  }
+  val SuffixRegex: Regex = """.+\.[^\.]*""".r
+  fileName match {
+    case SuffixRegex(_*) =>
+      Builder.warning(
+        s"""LoadMemoryAnnotation fileName "$fileName" has extension ".txt" will still be appended"""
+      )
+    case _ =>
+  }
```

Comment:
My problem here is that when the memory is an aggregate it puts a heavier burden on figuring out what the naming convention should be. My naive hope was that by the code auto adding the suffix we could keep it simpler.  if they control the suffix, what do you think I should do. Ask for a place holder like "memory_?.hex" where ? will be replaced by the memory subfield.

Reply:
You're absolutely right. Additionally, there's zero reason to try and provide spec compliance. We're in control of everything here and this will be presented to the user, assumedly, as an API for loading memory, _not an API for accessing readmemh/b Verilog primitives_. We could have the user pass a function in or something, but that smells over-engineered. I'm good with the `.txt` and your strategy here. Thanks for the explanation.